aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthSession.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.auth.cognito;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.amplifyframework.auth.AuthSession;
+import com.amplifyframework.auth.AuthUserState;
+
+import com.amazonaws.auth.AWSCredentials;
+
+import java.util.Objects;
+
+public final class AWSCognitoAuthSession extends AuthSession {
+    private AWSCredentials awsCredentials;
+    private String identityId;
+    private String accessToken;
+    private String idToken;
+    private String refreshToken;
+
+    public AWSCognitoAuthSession(
+            @NonNull AuthUserState state
+    ) {
+        super(state);
+    }
+
+    public AWSCognitoAuthSession(
+            @NonNull AuthUserState state,
+            @Nullable AWSCredentials awsCredentials,
+            @Nullable String identityId,
+            @Nullable String accessToken,
+            @Nullable String idToken,
+            @Nullable String refreshToken
+    ) {
+        super(Objects.requireNonNull(state));
+
+        this.awsCredentials = awsCredentials;
+        this.identityId = identityId;
+        this.accessToken = accessToken;
+        this.idToken = idToken;
+        this.refreshToken = refreshToken;
+    }
+
+    public AWSCredentials getAwsCredentials() {
+        return awsCredentials;
+    }
+
+    public String getIdentityId() {
+        return identityId;
+    }
+
+    public String getAccessToken() {
+        return accessToken;
+    }
+
+    public String getIdToken() {
+        return idToken;
+    }
+
+    public String getRefreshToken() {
+        return refreshToken;
+    }
+
+    public void setAwsCredentials(AWSCredentials awsCredentials) {
```

Comment:
Instead of mutation the session object, we might like to decouple these token/credentials.

You can "defer" the resolution of these things by using a provider delegate.

```
final class ImmutableSession {
    ImmutableSession(final Provider<AWSCredentials> credentialsProvider, final Provider<String> identityIdProvider) {
    }
```

This way, the session object isn't changing. But, the functions that are passed in to resolve fields may return different results.

```
new WhateverSession(/* provider */ () -> awsCredentials, /* provider */ () -> identityId);
```
Or maybe there's some implementation that's real crazy, not just returning a variable.

This is the pattern of the AWS credentials provider.